File: SpringAziron/src/main/java/management/pipeline/pipeNodes/skillPipeNode/SkillPipeNode.java
package management.pipeline.pipeNodes.skillPipeNode;

import heroes.abstractHero.hero.Hero;
import management.pipeline.APipeline;
import management.pipeline.pipeNodes.defaultPipeNode.DefaultPipeNode;
import management.playerManagement.PlayerManager;

public final class SkillPipeNode extends DefaultPipeNode {

    public SkillPipeNode(Hero hero, PlayerManager playerManager, APipeline pipeline) {
        super(hero, playerManager, pipeline);
    }

    @Override
    public String getPipeNodeID() {
        return null;
    }
}

File: SpringAziron/src/main/java/management/pipeline/pipeNodes/corePipeNodes/possibilityPipeNodes/skillPipeNode/SkillPipeNode.kt
package management.pipeline.pipeNodes.corePipeNodes.possibilityPipeNodes.skillPipeNode

import heroes.abstractHero.hero.Hero
import heroes.abstractHero.possibility.APossibility
import heroes.abstractHero.skills.Skill
import management.actionManagement.actions.ActionType
import management.pipeline.APipeline
import management.pipeline.pipeNodes.corePipeNodes.CorePipeNode
import management.pipeline.pipeNodes.corePipeNodes.possibilityPipeNodes.PossibilityPipeNode
import management.playerManagement.PlayerManager

class SkillPipeNode(pipeNodeID: String, hero: Hero, playerManager: PlayerManager, pipeline: APipeline)
    : CorePipeNode(pipeNodeID, hero, playerManager, pipeline) {

    init {
        this.engineComponentList.add(CoreBeforeUsedSkillEngineComponent(this.hero, this.pipeline, this.playerManager))
    }

    //Always last in engine order component:
    class CoreBeforeUsedSkillEngineComponent(hero: Hero, pipeline: APipeline, playerManager: PlayerManager)
        : PossibilityPipeNode.CoreBeforeUsedPossibilityEventEngine("CoreBeforeUsedSkillEngineComponent", hero
            , pipeline, playerManager) {

        override fun getCheckedBeforeActionType(): ActionType = ActionType.BEFORE_USED_SKILL

        override fun getActionEventDataFieldClass(): Class<out APossibility>  = Skill::class.java

        override fun getDuringActionType(): ActionType = ActionType.DURING_USING_SKILL
    }
}